File: LeetCode-Contests/LeetCode-contest-234/5716.cc
class Solution {
public:
  int maxNiceDivisors(int primeFactors) {
    unsigned long long m = 0;
    return decode(primeFactors, 3);
  }

private:
  vector<vector<int>> memo;
  unsigned long long decode(int num, int n) {
    if (num == 0)
      return 1;
    if(memo[num][n] >= 0) return memo[num][n];
    return memo[num][n] = decode(num - num / n, n - 1) * (num / n);
  }
};

File: LeetCode-Contests/LeetCode-contest-234/5716.py
class Solution:
    def maxNiceDivisors(self, primeFactors: int) -> int:
        return max([i for i in range(1, primeFactors)], key=lambda x: self.decode(primeFactors, i))
        return max([self.decode(primeFactors, i) for i in range(1, primeFactors+1)]) % (10**9 + 7)

    def decode(self, num, n):
        if num == 0:
            return 1
        return self.decode(num - num // n, n - 1) * num // n
